Change Sequence to use VarHandles instead of Unsafe

Palmr opened the pull request:
As part of moving towards JDK11 we should be removing uses of `sun.misc.Unsafe`
One of the main uses is `Sequence` and we can replace use of `Unsafe` with the newer `VarHandle` API.

The other use of `Unsafe` is in the `RingBuffer` class, I have not touched that as part of this change. I think that requires its own work and PR as its use of Unsafe is slightly different.

For benchmarking and stress testing I created three variants:

 - `SequenceUnsafe` - sequence as it was in Disruptor 3.4.2
 - `SequenceUnsafeVarHandle` - a version using VarHandles that was on the jdk9 branch
 - `SequenceUnsafeVarHandleBarrier` - We thought we could improve upon the basic use of VarHandles (**This is the variant I have pulled in to the main module**)

I ran all variations through the same jcstress tests I wrote for the Sequence as it was before I touched it. All tests pass.

I have also run the JMH benchmark on my local machine (no tuning so error fairly large) getting the following results:

```
$ cat /proc/cpuinfo | grep 'model name' | uniq
model name      : Intel(R) Core(TM) i5-6600K CPU @ 3.50GHz

# JMH version: 1.26
# VM version: JDK 11.0.8, OpenJDK 64-Bit Server VM, 11.0.8+10-LTS
# VM invoker: /home/nick/.jdks/corretto-11.0.8/bin/java
# VM options: -Xmx256m -Xms256m -XX:MaxDirectMemorySize=1g -Dfile.encoding=UTF-8 -Duser.country=GB -Duser.language=en -Duser.variant

Benchmark                                                      Mode   Cnt     Score     Error   Units
SequenceBenchmark.AtomicLong                                  thrpt    10  1627.474 ± 346.149  ops/us
SequenceBenchmark.SequenceUnsafe                              thrpt    10  1670.924 ± 375.118  ops/us
SequenceBenchmark.SequenceVarHandle                           thrpt    10  1384.866 ± 211.633  ops/us
SequenceBenchmark.SequenceVarHandleBarrier                    thrpt    10  1507.071 ± 353.475  ops/us

SequenceBenchmark.AtomicLong:incrementValue2                  thrpt    10    34.129 ±  33.359  ops/us
SequenceBenchmark.SequenceUnsafe:incrementValue2              thrpt    10    41.021 ±  17.756  ops/us
SequenceBenchmark.SequenceVarHandle:incrementValue2           thrpt    10    25.945 ±  11.946  ops/us
SequenceBenchmark.SequenceVarHandleBarrier:incrementValue2    thrpt    10    32.693 ±  14.031  ops/us

SequenceBenchmark.AtomicLong:read1                            thrpt    10    63.318 ±  34.628  ops/us
SequenceBenchmark.SequenceUnsafe:read1                        thrpt    10    78.373 ±  31.684  ops/us
SequenceBenchmark.SequenceVarHandle:read1                     thrpt    10    17.110 ±   4.096  ops/us
SequenceBenchmark.SequenceVarHandleBarrier:read1              thrpt    10    76.520 ±  19.497  ops/us

SequenceBenchmark.AtomicLong:read2                            thrpt    10   112.175 ± 116.361  ops/us
SequenceBenchmark.SequenceUnsafe:read2                        thrpt    10   179.929 ±  68.037  ops/us
SequenceBenchmark.SequenceVarHandle:read2                     thrpt    10    21.812 ±   9.321  ops/us
SequenceBenchmark.SequenceVarHandleBarrier:read2              thrpt    10   137.503 ±  62.813  ops/us

SequenceBenchmark.AtomicLong:setValue1Opaque                  thrpt    10  1390.576 ± 255.523  ops/us
SequenceBenchmark.SequenceUnsafe:setValue1                    thrpt    10  1335.698 ± 349.912  ops/us
SequenceBenchmark.SequenceVarHandle:setValue1                 thrpt    10  1287.291 ± 217.978  ops/us
SequenceBenchmark.SequenceVarHandleBarrier:setValue1          thrpt    10  1221.826 ± 338.011  ops/us

SequenceBenchmark.AtomicLong:setValue1Volatile                thrpt    10    27.276 ±  11.290  ops/us
SequenceBenchmark.SequenceUnsafe:setValue1Volatile            thrpt    10    35.903 ±  11.146  ops/us
SequenceBenchmark.SequenceVarHandle:setValue1Volatile         thrpt    10    32.708 ±   8.105  ops/us
SequenceBenchmark.SequenceVarHandleBarrier:setValue1Volatile  thrpt    10    38.528 ±   9.951  ops/us
```

This appears to show that the `SequenceVarHandleBarrier` has equivilent results to the old `SequenceUnsafe` methods, or at least well within error bounds from my local run.

Palmr commented:
Some changes I would like to make, but haven't put the time in yet:

 - Consolidate the alternatives package between jmh/jcstress
 - Run JMH on a tuned environment to get cleaner numbers
 - Run the old perf tests
 - There's a `FixedSequenceGroup` that I find slightly questionably. It overrides `set` so I can't have the constructor of `Sequence` call that to set the initial value. Perhaps I could do something there to stop the duplicated setting method between constructor and `Sequence::set()`. Does `FixedSequenceGroup` need to extend `Sequence`? Might a split reader/writer interface get us away from a class extending it only to throw errors on anything but `get`?

Palmr commented:
I thought I would run some of the legacy perf tests to check that the disruptor wasn't going to suffer from this change at some level higher than the JMH benchmark was testing.

I chose three of the sequenced throughput perf tests thinking they would give a good idea of the affects, if any, from this change.

The results below seem to indicate the barrier-based VarHandle version of `Sequence` gives comparable, or slightly improved, results.

```
$ ./gradlew clean perfJar
```

## OneToOneSequencedThroughputTest

`$ java -cp build/libs/disruptor-perf-*.jar com.lmax.disruptor.sequenced.OneToOneSequencedThroughputTest`

### SequenceUnsafe

```
Run 0, Disruptor=57,208,237 ops/sec BatchPercent=60.94% AverageBatchSize=2
Run 1, Disruptor=56,689,342 ops/sec BatchPercent=64.73% AverageBatchSize=2
Run 2, Disruptor=32,679,738 ops/sec BatchPercent=43.03% AverageBatchSize=1
Run 3, Disruptor=35,688,793 ops/sec BatchPercent=52.95% AverageBatchSize=2
Run 4, Disruptor=48,851,978 ops/sec BatchPercent=64.91% AverageBatchSize=2
Run 5, Disruptor=47,036,688 ops/sec BatchPercent=61.08% AverageBatchSize=2
Run 6, Disruptor=44,189,129 ops/sec BatchPercent=56.86% AverageBatchSize=2
```

### SequenceVarHandleBarrier

```
Run 0, Disruptor=68,073,519 ops/sec BatchPercent=70.66% AverageBatchSize=3
Run 1, Disruptor=60,790,273 ops/sec BatchPercent=63.97% AverageBatchSize=2
Run 2, Disruptor=72,098,053 ops/sec BatchPercent=73.09% AverageBatchSize=3
Run 3, Disruptor=71,377,587 ops/sec BatchPercent=72.06% AverageBatchSize=3
Run 4, Disruptor=72,358,900 ops/sec BatchPercent=72.88% AverageBatchSize=3
Run 5, Disruptor=76,745,970 ops/sec BatchPercent=74.48% AverageBatchSize=3
Run 6, Disruptor=75,528,700 ops/sec BatchPercent=73.84% AverageBatchSize=3
```

## ThreeToOneSequencedThroughputTest

`$ java -cp build/libs/disruptor-perf-*.jar com.lmax.disruptor.sequenced.ThreeToOneSequencedThroughputTest`

### SequenceUnsafe

```
Run 0, Disruptor=14,214,641 ops/sec BatchPercent=37.92% AverageBatchSize=1
Run 1, Disruptor=12,763,241 ops/sec BatchPercent=38.40% AverageBatchSize=1
Run 2, Disruptor=14,566,642 ops/sec BatchPercent=35.01% AverageBatchSize=1
Run 3, Disruptor=14,124,293 ops/sec BatchPercent=32.43% AverageBatchSize=1
Run 4, Disruptor=14,255,167 ops/sec BatchPercent=33.71% AverageBatchSize=1
Run 5, Disruptor=14,234,875 ops/sec BatchPercent=36.20% AverageBatchSize=1
Run 6, Disruptor=14,492,753 ops/sec BatchPercent=36.65% AverageBatchSize=1
```

### SequenceVarHandleBarrier

```
Run 0, Disruptor=14,992,503 ops/sec BatchPercent=35.88% AverageBatchSize=1
Run 1, Disruptor=13,956,734 ops/sec BatchPercent=39.10% AverageBatchSize=1
Run 2, Disruptor=15,491,866 ops/sec BatchPercent=38.06% AverageBatchSize=1
Run 3, Disruptor=14,556,040 ops/sec BatchPercent=27.33% AverageBatchSize=1
Run 4, Disruptor=14,914,243 ops/sec BatchPercent=30.49% AverageBatchSize=1
Run 5, Disruptor=15,243,902 ops/sec BatchPercent=33.89% AverageBatchSize=1
Run 6, Disruptor=14,880,952 ops/sec BatchPercent=30.98% AverageBatchSize=1
```

## OneToThreePipelineSequencedThroughputTest

`$ java -cp build/libs/disruptor-perf-*.jar com.lmax.disruptor.sequenced.OneToThreePipelineSequencedThroughputTest`

### SequenceUnsafe

```
Run 0, Disruptor=18,155,410 ops/sec BatchPercent=0.00% AverageBatchSize=-1
Run 1, Disruptor=19,924,287 ops/sec BatchPercent=0.00% AverageBatchSize=-1
Run 2, Disruptor=28,612,303 ops/sec BatchPercent=0.00% AverageBatchSize=-1
Run 3, Disruptor=28,224,668 ops/sec BatchPercent=0.00% AverageBatchSize=-1
Run 4, Disruptor=26,881,720 ops/sec BatchPercent=0.00% AverageBatchSize=-1
Run 5, Disruptor=27,578,599 ops/sec BatchPercent=0.00% AverageBatchSize=-1
Run 6, Disruptor=26,226,068 ops/sec BatchPercent=0.00% AverageBatchSize=-1
```

### SequenceVarHandleBarrier

```
Run 0, Disruptor=24,067,388 ops/sec BatchPercent=0.00% AverageBatchSize=-1
Run 1, Disruptor=28,449,502 ops/sec BatchPercent=0.00% AverageBatchSize=-1
Run 2, Disruptor=26,532,236 ops/sec BatchPercent=0.00% AverageBatchSize=-1
Run 3, Disruptor=29,334,115 ops/sec BatchPercent=0.00% AverageBatchSize=-1
Run 4, Disruptor=25,214,321 ops/sec BatchPercent=0.00% AverageBatchSize=-1
Run 5, Disruptor=28,645,087 ops/sec BatchPercent=0.00% AverageBatchSize=-1
Run 6, Disruptor=25,484,199 ops/sec BatchPercent=0.00% AverageBatchSize=-1
```

eliquinox commented:
I was curious to see how different sequence implementations compare, so I ran `SequenceBenchmark`. Here are the results:
```
Benchmark                                                      Mode  Cnt    Score    Error   Units
SequenceBenchmark.AtomicLong                                  thrpt   10  309.645 ± 48.025  ops/us
SequenceBenchmark.AtomicLong:incrementValue2                  thrpt   10    6.637 ±  3.260  ops/us
SequenceBenchmark.AtomicLong:read1                            thrpt   10   19.376 ±  4.086  ops/us
SequenceBenchmark.AtomicLong:read2                            thrpt   10   21.275 ±  7.877  ops/us
SequenceBenchmark.AtomicLong:setValue1Opaque                  thrpt   10  253.640 ± 49.788  ops/us
SequenceBenchmark.AtomicLong:setValue1Volatile                thrpt   10    8.717 ±  2.474  ops/us
SequenceBenchmark.SequenceDoublePadded                        thrpt   10  357.470 ± 34.567  ops/us
SequenceBenchmark.SequenceDoublePadded:incrementValue2        thrpt   10   14.879 ±  5.032  ops/us
SequenceBenchmark.SequenceDoublePadded:read1                  thrpt   10   31.300 ±  8.017  ops/us
SequenceBenchmark.SequenceDoublePadded:read2                  thrpt   10   42.848 ± 13.172  ops/us
SequenceBenchmark.SequenceDoublePadded:setValue1              thrpt   10  251.457 ± 35.322  ops/us
SequenceBenchmark.SequenceDoublePadded:setValue1Volatile      thrpt   10   16.984 ±  6.966  ops/us
SequenceBenchmark.SequenceUnsafe                              thrpt   10  335.615 ± 29.608  ops/us
SequenceBenchmark.SequenceUnsafe:incrementValue2              thrpt   10   16.080 ±  6.585  ops/us
SequenceBenchmark.SequenceUnsafe:read1                        thrpt   10   28.083 ±  6.487  ops/us
SequenceBenchmark.SequenceUnsafe:read2                        thrpt   10   42.476 ±  6.266  ops/us
SequenceBenchmark.SequenceUnsafe:setValue1                    thrpt   10  232.414 ± 35.509  ops/us
SequenceBenchmark.SequenceUnsafe:setValue1Volatile            thrpt   10   16.562 ±  5.460  ops/us
SequenceBenchmark.SequenceVarHandle                           thrpt   10  272.607 ± 35.278  ops/us
SequenceBenchmark.SequenceVarHandle:incrementValue2           thrpt   10    9.852 ±  2.280  ops/us
SequenceBenchmark.SequenceVarHandle:read1                     thrpt   10    5.688 ±  1.408  ops/us
SequenceBenchmark.SequenceVarHandle:read2                     thrpt   10    6.263 ±  1.587  ops/us
SequenceBenchmark.SequenceVarHandle:setValue1                 thrpt   10  235.635 ± 36.088  ops/us
SequenceBenchmark.SequenceVarHandle:setValue1Volatile         thrpt   10   15.170 ±  3.359  ops/us
SequenceBenchmark.SequenceVarHandleArray                      thrpt   10  173.557 ± 19.038  ops/us
SequenceBenchmark.SequenceVarHandleArray:incrementValue2      thrpt   10    8.901 ±  1.708  ops/us
SequenceBenchmark.SequenceVarHandleArray:read1                thrpt   10    5.209 ±  0.927  ops/us
SequenceBenchmark.SequenceVarHandleArray:read2                thrpt   10    6.003 ±  1.261  ops/us
SequenceBenchmark.SequenceVarHandleArray:setValue1            thrpt   10  140.074 ± 20.609  ops/us
SequenceBenchmark.SequenceVarHandleArray:setValue1Volatile    thrpt   10   13.370 ±  5.692  ops/us
SequenceBenchmark.SequenceVarHandleBarrier                    thrpt   10  376.806 ± 68.159  ops/us
SequenceBenchmark.SequenceVarHandleBarrier:incrementValue2    thrpt   10   10.834 ±  2.951  ops/us
SequenceBenchmark.SequenceVarHandleBarrier:read1              thrpt   10   24.977 ±  4.427  ops/us
SequenceBenchmark.SequenceVarHandleBarrier:read2              thrpt   10   44.981 ± 12.289  ops/us
SequenceBenchmark.SequenceVarHandleBarrier:setValue1          thrpt   10  280.967 ± 66.986  ops/us
SequenceBenchmark.SequenceVarHandleBarrier:setValue1Volatile  thrpt   10   15.047 ±  3.054  ops/us
``` 

What stood out for me is how well `SequenceDoublePadded` performed, though that being said, I am on Intel:
```
cat /proc/cpuinfo | grep 'model name' | uniq
model name      : Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz
```

Palmr commented:
@eliquinox `SequenceDoublePadded` was a test thrown in to check the suggestion of issue #231 
It should perform the same as the regular `SequenceUnsafe` as the logic is the same. If there was any advantage to be had from double-padding either side of the value and the argument of intel prefetchers defaulting to fetch two cache lines, I would expect to see a better benchmark results.

Your results seem consistent with mine: pretty much identical to single-padding. Which is why I should probably close that issue and why we're not going for double padding.

Something of note is that the prefetcher setting is a bios option, for me it was listed as "Adjacent Cache Line Prefetch". You may want to double check that was enabled for your run.

Palmr commented:
Just to double check my comments on double padding, I checked on my laptop with the following CPU:

```
$ cat /proc/cpuinfo | grep 'model name' | uniq
model name	: 11th Gen Intel(R) Core(TM) i7-1165G7 @ 2.80GHz
```

My laptop did not appear to expose a prefetcher adjacent cache line BIOS setting, that I can find at least, for Adjacent Cache Line Prefetch. Instead I went for reading/writing MSR values.
https://software.intel.com/content/dam/develop/external/us/en/documents-tps/325462-sdm-vol-1-2abcd-3abcd.pdf

Using this document I could see that the MSR for my CPU is `0x1A0` (page 4618) which has two bits, `19` and `37`.

Bit `19`
> Adjacent Cache Line Prefetch Disable (R/W)
When set to 1, the processor fetches the cache line that
contains data currently required by the processor. When
set to 0, the processor fetches cache lines that comprise a
cache line pair (128 bytes).

Bit `37`
> DCU Prefetcher Disable (R/W)
When set to 1, the DCU L1 data cache prefetcher is
disabled. The default value after reset is 0. BIOS may
write ‘1’ to disable this feature.
The DCU prefetcher is an L1 data cache prefetcher. When
the DCU prefetcher detects multiple loads from the same
line done within a time limit, the DCU prefetcher assumes
the next line will be required. The next line is prefetched
in to the L1 data cache from memory or L2.

I can check and change the state of these regardless of the BIOS settings using the [msr-tools](https://github.com/intel/msr-tools)
```
$ sudo rdmsr 0x1ad  -a -f 19:19
$ sudo rdmsr 0x1ad  -a -f 37:37
```

To enable/disable I used `$ sudo wrmsr 0x1ad <value>` where value was the whole MSR value with bits `19`/`37` set or unset.

With Adjacent Cache Line Prefetch and DCU Prefetcher **disabled**:

```
Benchmark                                                  Mode  Cnt     Score     Error   Units
SequenceBenchmark.SequenceDoublePadded                    thrpt   10  1401.858 ± 616.853  ops/us
SequenceBenchmark.SequenceDoublePadded:incrementValue2    thrpt   10    23.179 ±  27.652  ops/us
SequenceBenchmark.SequenceDoublePadded:read1              thrpt   10    60.761 ±  43.585  ops/us
SequenceBenchmark.SequenceDoublePadded:read2              thrpt   10   130.411 ±  74.767  ops/us
SequenceBenchmark.SequenceDoublePadded:setValue1          thrpt   10  1165.226 ± 583.864  ops/us
SequenceBenchmark.SequenceDoublePadded:setValue1Volatile  thrpt   10    22.281 ±   6.578  ops/us

SequenceBenchmark.SequenceUnsafe                          thrpt   10  1306.529 ± 505.247  ops/us
SequenceBenchmark.SequenceUnsafe:incrementValue2          thrpt   10    20.269 ±   9.244  ops/us
SequenceBenchmark.SequenceUnsafe:read1                    thrpt   10    74.507 ±  52.033  ops/us
SequenceBenchmark.SequenceUnsafe:read2                    thrpt   10   140.199 ±  54.734  ops/us
SequenceBenchmark.SequenceUnsafe:setValue1                thrpt   10  1037.796 ± 463.507  ops/us
SequenceBenchmark.SequenceUnsafe:setValue1Volatile        thrpt   10    33.758 ±   8.643  ops/us
```

With Adjacent Cache Line Prefetch and DCU Prefetcher **enabled**:

```
Benchmark                                                  Mode  Cnt     Score     Error   Units
SequenceBenchmark.SequenceDoublePadded                    thrpt   10  1370.445 ± 513.440  ops/us
SequenceBenchmark.SequenceDoublePadded:incrementValue2    thrpt   10    17.410 ±   4.179  ops/us
SequenceBenchmark.SequenceDoublePadded:read1              thrpt   10    53.128 ±  56.051  ops/us
SequenceBenchmark.SequenceDoublePadded:read2              thrpt   10   139.192 ±  35.533  ops/us
SequenceBenchmark.SequenceDoublePadded:setValue1          thrpt   10  1134.929 ± 525.329  ops/us
SequenceBenchmark.SequenceDoublePadded:setValue1Volatile  thrpt   10    25.786 ±   8.338  ops/us

SequenceBenchmark.SequenceUnsafe                          thrpt   10  1351.774 ± 552.570  ops/us
SequenceBenchmark.SequenceUnsafe:incrementValue2          thrpt   10    24.148 ±  25.879  ops/us
SequenceBenchmark.SequenceUnsafe:read1                    thrpt   10    52.500 ±  31.887  ops/us
SequenceBenchmark.SequenceUnsafe:read2                    thrpt   10   136.321 ±  66.184  ops/us
SequenceBenchmark.SequenceUnsafe:setValue1                thrpt   10  1111.580 ± 523.656  ops/us
SequenceBenchmark.SequenceUnsafe:setValue1Volatile        thrpt   10    27.225 ±   9.403  ops/us
```

The error values are quite high, this is due to running on an un-tuned system and a laptop CPU.
But the results look about what I expect regarding the double padding and this benchmark. The prefetcher fetching an adjacent cache line has no impact on the double-padded Sequence compared to the single cache line padded version.

Perhaps we should also enable the `perf` profiler in JMH which can run `perf` and record extra stats while running the JMH benchmark, where `cache-misses` might be interesting.

eliquinox reviewed (approved):
My only remark is related the one related to `getAndAdd` method. After review of usages of this method in other OSS, I was not able to locate any after having reviewed top 5 most used maven artifacts that use dirsruptor. That being said, if we expose this publicly, still makes sense to support it to the best extent possible.

Palmr commented on src/main/java/com/lmax/disruptor/Sequence.java:
```diff
@@ -112,55 +100,79 @@ public void set(final long value)
+     * @param increment The value to add to the sequence.
+     * @return The value after the increment.
+     */
+    public long getAndAdd(final long increment)
```
That returns an object, which then needs casting back to the primitive long.
This hand-rolled version should be functionally the same, and seemed to be faster. See:

```
SequenceBenchmark.SequenceVarHandle:incrementValue2           thrpt    10    25.945 ±  11.946  ops/us
SequenceBenchmark.SequenceVarHandleBarrier:incrementValue2    thrpt    10    32.693 ±  14.031  ops/us
```

Palmr commented:
Attached the jmh output running on a quiet, fairly well tuned system. All looks good for the manual-barrier version of the VarHandle code.

Summarised and grouped:

```
Benchmark                                                      Mode  Cnt     Score     Error   Units
SequenceBenchmark.SequenceUnsafe                              thrpt   10   680.819 ± 139.177  ops/us
SequenceBenchmark.SequenceVarHandle                           thrpt   10   459.511 ±  92.724  ops/us
SequenceBenchmark.SequenceVarHandleBarrier                    thrpt   10   722.702 ± 217.289  ops/us

SequenceBenchmark.SequenceUnsafe:incrementValue2              thrpt   10    26.053 ±   8.046  ops/us
SequenceBenchmark.SequenceVarHandle:incrementValue2           thrpt   10    13.838 ±   2.874  ops/us
SequenceBenchmark.SequenceVarHandleBarrier:incrementValue2    thrpt   10    17.455 ±   5.329  ops/us

SequenceBenchmark.SequenceUnsafe:read1                        thrpt   10    44.790 ±  10.238  ops/us
SequenceBenchmark.SequenceVarHandle:read1                     thrpt   10     8.462 ±   2.055  ops/us
SequenceBenchmark.SequenceVarHandleBarrier:read1              thrpt   10    47.351 ±  16.133  ops/us

SequenceBenchmark.SequenceUnsafe:read2                        thrpt   10    71.330 ±  18.582  ops/us
SequenceBenchmark.SequenceVarHandle:read2                     thrpt   10     9.385 ±   1.253  ops/us
SequenceBenchmark.SequenceVarHandleBarrier:read2              thrpt   10    77.072 ±  23.103  ops/us

SequenceBenchmark.SequenceUnsafe:setValue1                    thrpt   10   509.089 ± 140.703  ops/us
SequenceBenchmark.SequenceVarHandle:setValue1                 thrpt   10   398.278 ±  92.584  ops/us
SequenceBenchmark.SequenceVarHandleBarrier:setValue1          thrpt   10   544.731 ± 210.477  ops/us

SequenceBenchmark.SequenceUnsafe:setValue1Volatile            thrpt   10    29.558 ±  12.316  ops/us
SequenceBenchmark.SequenceVarHandle:setValue1Volatile         thrpt   10    29.548 ±  11.000  ops/us
SequenceBenchmark.SequenceVarHandleBarrier:setValue1Volatile  thrpt   10    36.093 ±  21.302  ops/us
```

Full:

```
$ cat /proc/cpuinfo | grep 'model name' | uniq
model name	: Intel(R) Xeon(R) CPU E5-2687W v4 @ 3.00GHz

$ ./gradlew jmh
# JMH version: 1.26
# VM version: JDK 11.0.6, OpenJDK 64-Bit Server VM, 11.0.6+10-LTS
# VM invoker: /var/lib/jenkins/workspace/_beant/cache/jdk11/1.11.0.6_zulu11.37.17_ca-1/bin/java
# VM options: -Xmx256m -Xms256m -XX:MaxDirectMemorySize=1g -Dfile.encoding=UTF-8 -Duser.country=GB -Duser.language=en -Duser.variant

Benchmark                                                      Mode  Cnt     Score     Error   Units
SequenceBenchmark.AtomicLong                                  thrpt   10   758.610 ± 428.421  ops/us
SequenceBenchmark.AtomicLong:incrementValue2                  thrpt   10    34.829 ±  26.588  ops/us
SequenceBenchmark.AtomicLong:read1                            thrpt   10    53.325 ±  27.563  ops/us
SequenceBenchmark.AtomicLong:read2                            thrpt   10    87.316 ±  37.024  ops/us
SequenceBenchmark.AtomicLong:setValue1Opaque                  thrpt   10   557.494 ± 398.593  ops/us
SequenceBenchmark.AtomicLong:setValue1Volatile                thrpt   10    25.646 ±   9.353  ops/us
SequenceBenchmark.AtomicLong:·pauses                          thrpt    8   197.722                ms
SequenceBenchmark.AtomicLong:·pauses.avg                      thrpt         24.715                ms
SequenceBenchmark.AtomicLong:·pauses.count                    thrpt          8.000                 #
SequenceBenchmark.AtomicLong:·pauses.p0.00                    thrpt         11.010                ms
SequenceBenchmark.AtomicLong:·pauses.p0.50                    thrpt         26.755                ms
SequenceBenchmark.AtomicLong:·pauses.p0.90                    thrpt         32.768                ms
SequenceBenchmark.AtomicLong:·pauses.p0.95                    thrpt         32.768                ms
SequenceBenchmark.AtomicLong:·pauses.p0.99                    thrpt         32.768                ms
SequenceBenchmark.AtomicLong:·pauses.p0.999                   thrpt         32.768                ms
SequenceBenchmark.AtomicLong:·pauses.p0.9999                  thrpt         32.768                ms
SequenceBenchmark.AtomicLong:·pauses.p1.00                    thrpt         32.768                ms
SequenceBenchmark.SequenceDoublePadded                        thrpt   10   649.292 ± 190.164  ops/us
SequenceBenchmark.SequenceDoublePadded:incrementValue2        thrpt   10    28.402 ±   8.522  ops/us
SequenceBenchmark.SequenceDoublePadded:read1                  thrpt   10    48.758 ±  15.353  ops/us
SequenceBenchmark.SequenceDoublePadded:read2                  thrpt   10    73.083 ±  21.750  ops/us
SequenceBenchmark.SequenceDoublePadded:setValue1              thrpt   10   468.682 ± 187.049  ops/us
SequenceBenchmark.SequenceDoublePadded:setValue1Volatile      thrpt   10    30.366 ±   9.519  ops/us
SequenceBenchmark.SequenceDoublePadded:·pauses                thrpt   15   295.191                ms
SequenceBenchmark.SequenceDoublePadded:·pauses.avg            thrpt         19.679                ms
SequenceBenchmark.SequenceDoublePadded:·pauses.count          thrpt         15.000                 #
SequenceBenchmark.SequenceDoublePadded:·pauses.p0.00          thrpt         11.076                ms
SequenceBenchmark.SequenceDoublePadded:·pauses.p0.50          thrpt         19.857                ms
SequenceBenchmark.SequenceDoublePadded:·pauses.p0.90          thrpt         30.828                ms
SequenceBenchmark.SequenceDoublePadded:·pauses.p0.95          thrpt         36.766                ms
SequenceBenchmark.SequenceDoublePadded:·pauses.p0.99          thrpt         36.766                ms
SequenceBenchmark.SequenceDoublePadded:·pauses.p0.999         thrpt         36.766                ms
SequenceBenchmark.SequenceDoublePadded:·pauses.p0.9999        thrpt         36.766                ms
SequenceBenchmark.SequenceDoublePadded:·pauses.p1.00          thrpt         36.766                ms
SequenceBenchmark.SequenceUnsafe                              thrpt   10   680.819 ± 139.177  ops/us
SequenceBenchmark.SequenceUnsafe:incrementValue2              thrpt   10    26.053 ±   8.046  ops/us
SequenceBenchmark.SequenceUnsafe:read1                        thrpt   10    44.790 ±  10.238  ops/us
SequenceBenchmark.SequenceUnsafe:read2                        thrpt   10    71.330 ±  18.582  ops/us
SequenceBenchmark.SequenceUnsafe:setValue1                    thrpt   10   509.089 ± 140.703  ops/us
SequenceBenchmark.SequenceUnsafe:setValue1Volatile            thrpt   10    29.558 ±  12.316  ops/us
SequenceBenchmark.SequenceUnsafe:·pauses                      thrpt    8   163.758                ms
SequenceBenchmark.SequenceUnsafe:·pauses.avg                  thrpt         20.470                ms
SequenceBenchmark.SequenceUnsafe:·pauses.count                thrpt          8.000                 #
SequenceBenchmark.SequenceUnsafe:·pauses.p0.00                thrpt         12.665                ms
SequenceBenchmark.SequenceUnsafe:·pauses.p0.50                thrpt         17.818                ms
SequenceBenchmark.SequenceUnsafe:·pauses.p0.90                thrpt         30.409                ms
SequenceBenchmark.SequenceUnsafe:·pauses.p0.95                thrpt         30.409                ms
SequenceBenchmark.SequenceUnsafe:·pauses.p0.99                thrpt         30.409                ms
SequenceBenchmark.SequenceUnsafe:·pauses.p0.999               thrpt         30.409                ms
SequenceBenchmark.SequenceUnsafe:·pauses.p0.9999              thrpt         30.409                ms
SequenceBenchmark.SequenceUnsafe:·pauses.p1.00                thrpt         30.409                ms
SequenceBenchmark.SequenceVarHandle                           thrpt   10   459.511 ±  92.724  ops/us
SequenceBenchmark.SequenceVarHandle:incrementValue2           thrpt   10    13.838 ±   2.874  ops/us
SequenceBenchmark.SequenceVarHandle:read1                     thrpt   10     8.462 ±   2.055  ops/us
SequenceBenchmark.SequenceVarHandle:read2                     thrpt   10     9.385 ±   1.253  ops/us
SequenceBenchmark.SequenceVarHandle:setValue1                 thrpt   10   398.278 ±  92.584  ops/us
SequenceBenchmark.SequenceVarHandle:setValue1Volatile         thrpt   10    29.548 ±  11.000  ops/us
SequenceBenchmark.SequenceVarHandle:·pauses                   thrpt   44  1007.419                ms
SequenceBenchmark.SequenceVarHandle:·pauses.avg               thrpt         22.896                ms
SequenceBenchmark.SequenceVarHandle:·pauses.count             thrpt         44.000                 #
SequenceBenchmark.SequenceVarHandle:·pauses.p0.00             thrpt         11.911                ms
SequenceBenchmark.SequenceVarHandle:·pauses.p0.50             thrpt         23.347                ms
SequenceBenchmark.SequenceVarHandle:·pauses.p0.90             thrpt         32.244                ms
SequenceBenchmark.SequenceVarHandle:·pauses.p0.95             thrpt         34.521                ms
SequenceBenchmark.SequenceVarHandle:·pauses.p0.99             thrpt         35.717                ms
SequenceBenchmark.SequenceVarHandle:·pauses.p0.999            thrpt         35.717                ms
SequenceBenchmark.SequenceVarHandle:·pauses.p0.9999           thrpt         35.717                ms
SequenceBenchmark.SequenceVarHandle:·pauses.p1.00             thrpt         35.717                ms
SequenceBenchmark.SequenceVarHandleArray                      thrpt   10   320.805 ± 106.894  ops/us
SequenceBenchmark.SequenceVarHandleArray:incrementValue2      thrpt   10    12.294 ±   4.216  ops/us
SequenceBenchmark.SequenceVarHandleArray:read1                thrpt   10     8.217 ±   1.658  ops/us
SequenceBenchmark.SequenceVarHandleArray:read2                thrpt   10    11.447 ±   3.789  ops/us
SequenceBenchmark.SequenceVarHandleArray:setValue1            thrpt   10   262.273 ± 100.458  ops/us
SequenceBenchmark.SequenceVarHandleArray:setValue1Volatile    thrpt   10    26.575 ±   8.423  ops/us
SequenceBenchmark.SequenceVarHandleArray:·pauses              thrpt   50   946.594                ms
SequenceBenchmark.SequenceVarHandleArray:·pauses.avg          thrpt         18.932                ms
SequenceBenchmark.SequenceVarHandleArray:·pauses.count        thrpt         50.000                 #
SequenceBenchmark.SequenceVarHandleArray:·pauses.p0.00        thrpt          4.850                ms
SequenceBenchmark.SequenceVarHandleArray:·pauses.p0.50        thrpt         16.736                ms
SequenceBenchmark.SequenceVarHandleArray:·pauses.p0.90        thrpt         30.130                ms
SequenceBenchmark.SequenceVarHandleArray:·pauses.p0.95        thrpt         34.387                ms
SequenceBenchmark.SequenceVarHandleArray:·pauses.p0.99        thrpt         45.023                ms
SequenceBenchmark.SequenceVarHandleArray:·pauses.p0.999       thrpt         45.023                ms
SequenceBenchmark.SequenceVarHandleArray:·pauses.p0.9999      thrpt         45.023                ms
SequenceBenchmark.SequenceVarHandleArray:·pauses.p1.00        thrpt         45.023                ms
SequenceBenchmark.SequenceVarHandleBarrier                    thrpt   10   722.702 ± 217.289  ops/us
SequenceBenchmark.SequenceVarHandleBarrier:incrementValue2    thrpt   10    17.455 ±   5.329  ops/us
SequenceBenchmark.SequenceVarHandleBarrier:read1              thrpt   10    47.351 ±  16.133  ops/us
SequenceBenchmark.SequenceVarHandleBarrier:read2              thrpt   10    77.072 ±  23.103  ops/us
SequenceBenchmark.SequenceVarHandleBarrier:setValue1          thrpt   10   544.731 ± 210.477  ops/us
SequenceBenchmark.SequenceVarHandleBarrier:setValue1Volatile  thrpt   10    36.093 ±  21.302  ops/us
SequenceBenchmark.SequenceVarHandleBarrier:·pauses            thrpt    9   219.857                ms
SequenceBenchmark.SequenceVarHandleBarrier:·pauses.avg        thrpt         24.429                ms
SequenceBenchmark.SequenceVarHandleBarrier:·pauses.count      thrpt          9.000                 #
SequenceBenchmark.SequenceVarHandleBarrier:·pauses.p0.00      thrpt         12.452                ms
SequenceBenchmark.SequenceVarHandleBarrier:·pauses.p0.50      thrpt         25.854                ms
SequenceBenchmark.SequenceVarHandleBarrier:·pauses.p0.90      thrpt         42.533                ms
SequenceBenchmark.SequenceVarHandleBarrier:·pauses.p0.95      thrpt         42.533                ms
SequenceBenchmark.SequenceVarHandleBarrier:·pauses.p0.99      thrpt         42.533                ms
SequenceBenchmark.SequenceVarHandleBarrier:·pauses.p0.999     thrpt         42.533                ms
SequenceBenchmark.SequenceVarHandleBarrier:·pauses.p0.9999    thrpt         42.533                ms
SequenceBenchmark.SequenceVarHandleBarrier:·pauses.p1.00      thrpt         42.533                ms
BlockingQueueBenchmark.producing                               avgt    5   152.510 ±  72.382   ns/op
BlockingQueueBenchmark.producing:·pauses                       avgt   12   224.100                ms
BlockingQueueBenchmark.producing:·pauses.avg                   avgt         18.675                ms
BlockingQueueBenchmark.producing:·pauses.count                 avgt         12.000                 #
BlockingQueueBenchmark.producing:·pauses.p0.00                 avgt          8.962                ms
BlockingQueueBenchmark.producing:·pauses.p0.50                 avgt         17.809                ms
BlockingQueueBenchmark.producing:·pauses.p0.90                 avgt         31.837                ms
BlockingQueueBenchmark.producing:·pauses.p0.95                 avgt         33.686                ms
BlockingQueueBenchmark.producing:·pauses.p0.99                 avgt         33.686                ms
BlockingQueueBenchmark.producing:·pauses.p0.999                avgt         33.686                ms
BlockingQueueBenchmark.producing:·pauses.p0.9999               avgt         33.686                ms
BlockingQueueBenchmark.producing:·pauses.p1.00                 avgt         33.686                ms
SingleProducerSingleConsumer.producing                         avgt    5    16.673 ±   3.786   ns/op
SingleProducerSingleConsumer.producing:·pauses                 avgt    2    39.715                ms
SingleProducerSingleConsumer.producing:·pauses.avg             avgt         19.857                ms
SingleProducerSingleConsumer.producing:·pauses.count           avgt          2.000                 #
SingleProducerSingleConsumer.producing:·pauses.p0.00           avgt         17.760                ms
SingleProducerSingleConsumer.producing:·pauses.p0.50           avgt         19.857                ms
SingleProducerSingleConsumer.producing:·pauses.p0.90           avgt         21.955                ms
SingleProducerSingleConsumer.producing:·pauses.p0.95           avgt         21.955                ms
SingleProducerSingleConsumer.producing:·pauses.p0.99           avgt         21.955                ms
SingleProducerSingleConsumer.producing:·pauses.p0.999          avgt         21.955                ms
SingleProducerSingleConsumer.producing:·pauses.p0.9999         avgt         21.955                ms
SingleProducerSingleConsumer.producing:·pauses.p1.00           avgt         21.955                ms
```

Palmr commented:
@sea36 I have added some optional thread pinning code to the `SequenceBenchmark` which I think is well isolated from our regular codebase and appears to be functional. I have yet to try this on the same system as yesterday however.

I realised it wasn't using the thread state because I wasn't passing it in to my benchmarks!
I also thought two states were doubling the thread count, but it was actually the `main` in that class setting the thread count based on system core count. I have now removed that and annotated the class with a `@Threads(1)` which should make clear it should only be running one thread per benchmark in each group, aka 5 threads per group.

Palmr commented:
An initial run of `./graldew jscress` on an ARM64 CPU (Raspberry Pi 3 Model B, BCM2837 SoC with a 1.2 GHz 64-bit quad-core ARM Cortex-A53 processor, with 512 KiB shared L2 cache) showed some failures from `com.lmax.disruptor.SequenceStressVarHandleBarrier.SetVolatileGuard`.

Looking in to this, using `VarHandle.fullFence()` on its own is not enough to get full volatile semantics (http://gee.cs.oswego.edu/dl/html/j9mm.html#volatilefencesec)

I have added in a `VarHandle.releaseFence()` before the store, keeping the `fullFence()` after, and now jcstress seems happier.

I have checked the impact of this on x86 benchmarks and it continues to be good compared to the unsafe alternative and `VarHandle.setVolatile()` methods.

```
Benchmark                                                      Mode  Cnt     Score     Error   Units
SequenceBenchmark.SequenceUnsafe                              thrpt   10  1142.050 ± 650.798  ops/us
SequenceBenchmark.SequenceVarHandle                           thrpt   10  1115.075 ± 253.574  ops/us
SequenceBenchmark.SequenceVarHandleBarrier                    thrpt   10  1282.676 ± 476.582  ops/us

SequenceBenchmark.SequenceUnsafe:incrementValue2              thrpt   10    25.233 ±  26.006  ops/us
SequenceBenchmark.SequenceVarHandle:incrementValue2           thrpt   10    17.456 ±   8.243  ops/us
SequenceBenchmark.SequenceVarHandleBarrier:incrementValue2    thrpt   10    10.754 ±   4.297  ops/us

SequenceBenchmark.SequenceUnsafe:read1                        thrpt   10    64.086 ±  77.581  ops/us
SequenceBenchmark.SequenceVarHandle:read1                     thrpt   10    12.071 ±   4.090  ops/us
SequenceBenchmark.SequenceVarHandleBarrier:read1              thrpt   10    47.994 ±  39.471  ops/us

SequenceBenchmark.SequenceUnsafe:read2                        thrpt   10   129.671 ±  48.760  ops/us
SequenceBenchmark.SequenceVarHandle:read2                     thrpt   10    12.759 ±   5.144  ops/us
SequenceBenchmark.SequenceVarHandleBarrier:read2              thrpt   10   150.564 ±  66.268  ops/us

SequenceBenchmark.SequenceUnsafe:setValue1Volatile            thrpt   10    28.560 ±  10.771  ops/us
SequenceBenchmark.SequenceVarHandle:setValue1Volatile         thrpt   10    27.699 ±  11.473  ops/us
SequenceBenchmark.SequenceVarHandleBarrier:setValue1Volatile  thrpt   10    21.803 ±  13.035  ops/us

SequenceBenchmark.SequenceUnsafe:setValue1                    thrpt   10   894.500 ± 649.368  ops/us
SequenceBenchmark.SequenceVarHandle:setValue1                 thrpt   10  1045.091 ± 259.614  ops/us
SequenceBenchmark.SequenceVarHandleBarrier:setValue1          thrpt   10  1051.561 ± 531.837  ops/us
```

Palmr commented:
For anyone curious what the results are like running `$ ./gradlew clean jmh` on a Raspberry pi:

```
Benchmark                                                      Mode  Cnt    Score    Error   Units
SequenceBenchmark.AtomicLong                                  thrpt   10   63.240 ± 23.713  ops/us
SequenceBenchmark.SequenceUnsafe                              thrpt   10   76.656 ±  2.005  ops/us
SequenceBenchmark.SequenceVarHandle                           thrpt   10   54.037 ±  6.468  ops/us
SequenceBenchmark.SequenceVarHandleBarrier                    thrpt   10   74.333 ±  1.955  ops/us

SequenceBenchmark.AtomicLong:incrementValue2                  thrpt   10    5.827 ±  5.842  ops/us
SequenceBenchmark.SequenceUnsafe:incrementValue2              thrpt   10    9.831 ±  2.603  ops/us
SequenceBenchmark.SequenceVarHandle:incrementValue2           thrpt   10    3.718 ±  1.072  ops/us
SequenceBenchmark.SequenceVarHandleBarrier:incrementValue2    thrpt   10   10.435 ±  2.181  ops/us

SequenceBenchmark.AtomicLong:read1                            thrpt   10   10.743 ±  2.442  ops/us
SequenceBenchmark.SequenceUnsafe:read1                        thrpt   10   11.671 ±  1.056  ops/us
SequenceBenchmark.SequenceVarHandle:read1                     thrpt   10    2.152 ±  0.693  ops/us
SequenceBenchmark.SequenceVarHandleBarrier:read1              thrpt   10   11.604 ±  1.750  ops/us

SequenceBenchmark.AtomicLong:read2                            thrpt   10   11.749 ±  3.824  ops/us
SequenceBenchmark.SequenceUnsafe:read2                        thrpt   10   12.641 ±  2.604  ops/us
SequenceBenchmark.SequenceVarHandle:read2                     thrpt   10    2.440 ±  0.448  ops/us
SequenceBenchmark.SequenceVarHandleBarrier:read2              thrpt   10   13.959 ±  2.870  ops/us

SequenceBenchmark.AtomicLong:setValue1Opaque                  thrpt   10   21.304 ± 10.284  ops/us
SequenceBenchmark.SequenceUnsafe:setValue1                    thrpt   10   23.914 ±  4.512  ops/us
SequenceBenchmark.SequenceVarHandle:setValue1                 thrpt   10   24.894 ± 10.429  ops/us
SequenceBenchmark.SequenceVarHandleBarrier:setValue1          thrpt   10   19.299 ±  4.137  ops/us

SequenceBenchmark.AtomicLong:setValue1Volatile                thrpt   10   13.616 ±  4.653  ops/us
SequenceBenchmark.SequenceUnsafe:setValue1Volatile            thrpt   10   18.599 ±  3.102  ops/us
SequenceBenchmark.SequenceVarHandle:setValue1Volatile         thrpt   10   20.833 ±  7.153  ops/us
SequenceBenchmark.SequenceVarHandleBarrier:setValue1Volatile  thrpt   10   19.037 ±  4.829  ops/us
```

Palmr commented:
Results from a tuned system with thread pinning to reduce error:

```
$ cat /proc/cpuinfo | grep 'model name' | uniq
model name	: Intel(R) Xeon(R) CPU E5-2687W v4 @ 3.00GHz

$ <insert tangle of cset, taskset and java> ./gradlew jmh

Benchmark                                                      Mode  Cnt     Score     Error   Units
SequenceBenchmark.SequenceUnsafe                              thrpt   10  1792.810 ±  24.144  ops/us
SequenceBenchmark.SequenceVarHandle                           thrpt   10  1663.844 ±   7.037  ops/us
SequenceBenchmark.SequenceVarHandleBarrier                    thrpt   10  1770.824 ±  26.133  ops/us

SequenceBenchmark.SequenceUnsafe:incrementValue2              thrpt   10    26.814 ±   2.853  ops/us
SequenceBenchmark.SequenceVarHandle:incrementValue2           thrpt   10    29.134 ±   3.272  ops/us
SequenceBenchmark.SequenceVarHandleBarrier:incrementValue2    thrpt   10    15.689 ±   3.561  ops/us

SequenceBenchmark.SequenceUnsafe:read1                        thrpt   10    70.707 ±  10.560  ops/us
SequenceBenchmark.SequenceVarHandle:read1                     thrpt   10     7.558 ±   3.078  ops/us
SequenceBenchmark.SequenceVarHandleBarrier:read1              thrpt   10    50.950 ±   4.342  ops/us

SequenceBenchmark.SequenceUnsafe:read2                        thrpt   10    85.321 ±  15.839  ops/us
SequenceBenchmark.SequenceVarHandle:read2                     thrpt   10    11.887 ±   1.084  ops/us
SequenceBenchmark.SequenceVarHandleBarrier:read2              thrpt   10    89.497 ±  18.699  ops/us

SequenceBenchmark.SequenceUnsafe:setValue1                    thrpt   10  1586.187 ±   4.585  ops/us
SequenceBenchmark.SequenceVarHandle:setValue1                 thrpt   10  1582.842 ±   2.760  ops/us
SequenceBenchmark.SequenceVarHandleBarrier:setValue1          thrpt   10  1586.548 ±   1.385  ops/us

SequenceBenchmark.SequenceUnsafe:setValue1Volatile            thrpt   10    23.780 ±   1.363  ops/us
SequenceBenchmark.SequenceVarHandle:setValue1Volatile         thrpt   10    32.421 ±   5.780  ops/us
SequenceBenchmark.SequenceVarHandleBarrier:setValue1Volatile  thrpt   10    28.141 ±   4.156  ops/us
```

This shows `SequenceVarHandleBarrier:incrementValue2` looking slower than the other implementations. This method is not really used however, perhaps we could speed it up, deprecate it, or both?

I also noted 1 gc happened while benchmarking the barrier version. This may be worth looking in to more.

grumpyjames reviewed (approved):
I think most of this is non-controversial; we're mostly replacing the existing Sequence with the VarHandleBarrier version, which you additionally found a bug in and fixed (great job trying it on a pi!).

I think we should merge this, and then go see how much of the slow/misnamed parts we can delete (we should be able to skip deprecation if we're issuing a v4 into the world, right?)

sea36 commented:
I don't have anything to add to what has already been said.

We should go ahead and merge.

I agree - switch to v4, migration to Java 11, is big enough already that we can easily make small breaking changes to the API.
